File: platform/core/messaging/src/main/kotlin/org/whiteprint/platform/core/messaging/outbox/EventOutbox.kt
package org.whiteprint.platform.core.messaging.outbox

import org.whiteprint.platform.core.messaging.model.event.EventScope
import org.whiteprint.platform.core.messaging.model.event.EventStatus
import java.time.Instant

interface  EventOutbox {

    val eventId: Long

    val traceId: String

    val causationId: String?

    val occurredAt: Instant

    val issuer: String

    val producer: String

    val schemaVersion: String

    val partitionKey: Long

    val eventScope: EventScope

    val eventType: String

    val payload: ByteArray

    val payloadJson: String

    val metadataJson: String

    val status: EventStatus

    val attemptCount: Int

    val lastAttemptedAt: Instant?

}

File: platform/adapter/event/outbox/src/main/resources/mongo/eventOutbox.js
// Collection: event_outbox
// Document structure reference

db.createCollection("event_outbox");

db.event_outbox.createIndex({ status: 1, occurred_at: 1 });

// Document example:
// {
//     _id:            Long,           // event_id
//     trace_id:       String,
//     causation_id:   String | null,
//     occurred_at:    Date,
//     issuer:         String,
//     producer:       String,
//     schema_version: String,
//     partition_key:  Long,
//     event_scope:    String,         // EventScope enum
//     event_type:     String,
//     payload:        BinData,
//     payload_json:   String,
//     metadata_json:  String,
//     status:         String,         // PENDING | PROCESSING | PUBLISHED | FAILED
//     attempt_count:  Int,
//     last_attempted_at: Date | null
// }
